File: count_k_primes.cpp
/** ===========================================================================================
* Name : count_k_primes.cpp
* Author: taikido
* Description: Count K-Primes
* Source: http://www.codechef.com/problems/KPRIME
* Date: Fri.Dec.21.2013
*=============================================================================================**/

#include <iostream>
#include <cstdio>

using namespace std;

#define max_num 100001
#define max_k  6


void count_primes(int num_lines){
	bool is_prime[max_num];
	int counts[max_num];
	int kp[max_num][max_k];


	/*Initializations*/
	for (int i=0; i<max_num; i++) is_prime[i] = true;
	for (int i=0; i<max_num; i++) counts[i] = 0;


	for (int p=2; p < max_num; p++){
		if (is_prime[p]){
			counts[p] = 1;
			int j = 2;

			while ((p* j) < max_num){
				is_prime[p*j] = 0;
				counts[p*j]++;
				j++;
			}
		}
	}


	//Create running sum of the counts for the unique prime factors in each num from 2 -> max_num
	for (int i=2; i < max_num; i++){
		for(int j=1; j < max_k; j++)
			kp[i][j] = kp[i-1][j] + (counts[i] == j ? 1 : 0);
	}

	int a, b, k;
	for (int i=0; i < num_lines; i++){
		scanf("%d%d%d",&a,&b,&k);
		printf("%d\n",kp[b][k]-kp[a-1][k]);
	}

}



int main(){
	int t;
	scanf("%d",&t);
	count_primes(t);
	return 0;
}

File: easy/count_k_primes.py
"""===========================================================================================
* Name : count_k_primes.py
* Author: taikido
* Description: Count K-Primes
* Source: http://www.codechef.com/problems/KPRIME
* Date: Fri.Dec.21.2013
*============================================================================================="""

import sys
max_num = 100001


def count_primes(num_lines):
    is_prime = [1] * max_num
    counts = [0] * max_num
    
    #print "is_prime: ", is_prime

    for p in range(2, max_num):
        if is_prime[p]:
            counts[p] = 1
            j = 2
            while p * j < max_num:
                #print "p: {0}, j: {1}".format(p, j)
                is_prime[p*j] = 0
                counts[p*j] += 1
                j += 1
                
    kp = [[0 for i in range(6)] for j in range(max_num)]
    
    #print "counts: ", counts
    
    #Create running sum of the counts for the unique prime factors in each num from 2 -> max_num
    for i in range(2, max_num):
        for j in range(1, 6):
            kp[i][j] = kp[i-1][j] + (1 if counts[i] == j else 0)
    
    #print "kp: ", kp      
    for i in range(num_lines):
        a, b, k = map(int, sys.stdin.readline().strip().split())
        print kp[b][k] - kp[a-1][k]   


    

s = raw_input()
num_lines = map(int, s.split())[0]            
count_primes(num_lines)
 




